optuna/samplers/grid.py

```diff
@@ -0,0 +1,177 @@
+        # to validate the sampled value.
+
+        unvisited_grids = self._get_unvisited_grid_ids(study)
+
+        if len(unvisited_grids) == 0:
+            raise ValueError('All grids have been evaluated. If you want to avoid this error, '
+                             'please make sure that unnecessary trials do not run during '
+                             'optimization by properly setting `n_trials` in `study.optimize`.')
+
+        # In distributed optimization, multiple workers may simultaneously pick up the same grid.
+        # To make the conflict less frequent, the grid is chosen randomly.
+        grid_id = random.choice(unvisited_grids)
+
+        study._storage.set_trial_system_attr(trial._trial_id, 'search_space', self._search_space)
+        study._storage.set_trial_system_attr(trial._trial_id, 'grid_id', grid_id)
+
+        return {}
+
+    def sample_independent(self, study, trial, param_name, param_distribution):
+        # type: (Study, FrozenTrial, str, BaseDistribution) -> Any
+
+        if param_name not in self._search_space:
+            message = 'The parameter name, {}, is not found in the given grid.'.format(param_name)
+            raise ValueError(message)
+
+        grid_id = trial.system_attrs['grid_id']
+        param_value = self._all_grids[grid_id][self._param_names.index(param_name)]
+        contains = param_distribution._contains(param_distribution.to_internal_repr(param_value))
+        if not contains:
+            raise ValueError('The value `{}` is out of range of the parameter `{}`. Please make '
+                             'sure the search space of the `GridSampler` only contains values '
+                             'consistent with the distribution specified in the objective '
+                             'function. The distribution is: `{}`.'
+                             .format(param_value, param_name, param_distribution))
+
+        return param_value
+
+    @staticmethod
+    def _check_value(param_name, param_value):
+        # type: (str, Any) -> None
+
+        if param_value is None or isinstance(param_value, (str, int, float, bool)):
+            return
+
+        raise ValueError('{} contains a value with the type of {}, which is not supported by '
+                         '`GridSampler`. Please make sure a value is `str`, `int`, `float`, `bool`'
+                         ' or `None`.'.format(param_name, type(param_value)))
+
+    def _get_unvisited_grid_ids(self, study):
+        # type: (Study) -> List[int]
+
+        # List up all finished trials in the same search space.
+        trials = study.trials
+        trials = [t for t in trials if t.state.is_finished()]
+        trials = [t for t in trials if 'grid_id' in t.system_attrs]
+        trials = [t for t in trials if self._same_search_space(t.system_attrs['search_space'])]
+
+        # List up unvisited trials based on already finished ones.
+        visited_grids = [t.system_attrs['grid_id'] for t in trials]
```

Comment:
I believe you could construct `visited_grids` with a single loop over all trials (instead of 4). What do you think?

Reply:
Replaced the logic with a single for loop.